File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        if (k < 0) return 0; // absolute difference can't be negative
        
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int l = 0, r = 1;
        int count = 0;
        
        while (r < n && l < n) {
            if (l == r || nums[r] - nums[l] < k) {
                r++;
            } else if (nums[r] - nums[l] > k) {
                l++;
            } else {
                count++;
                l++;
                r++;
               
                while (r < n && nums[r] == nums[r - 1]) r++;
            }
        }
        
        return count;
    }
};

File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.java
class Solution {
    static {
        for (int i = 0; i < 500; i++) {
            findPairs(new int[] { -1, 0 }, 0);
        }
    }

    public static int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i : nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        int cnt = 0;
        for (int i : map.keySet()) {
            if (k == 0) {
                if (map.get(i) > 1)
                    cnt++;
            } else {
                if (map.containsKey(i - k))
                    cnt++;
            }
        }
        return cnt;
    }
}